eio/tls_eio.ml

```diff
@@ -5,198 +5,143 @@ exception Tls_failure of Tls.Engine.failure
+
   let rec drain_handshake t =
-    let push_linger t mcs =
-      match (mcs, t.linger) with
-      | (None, _)         -> ()
-      | (scs, None)       -> t.linger <- scs
-      | (Some cs, Some l) -> t.linger <- Some (Cstruct.append l cs)
-    in
-    match t.state with
-    | `Active tls when not (Tls.Engine.handshake_in_progress tls) ->
-        t
-    | _ ->
-        let cs = read_react t in
-        push_linger t cs; drain_handshake t
+    if not (Tls.Engine.handshake_in_progress t.tls) then t
+    else
+      let application_data = read_t t in
+      Option.iter (fun data ->
+          t.linger <-
+            if Cstruct.is_empty t.linger then data
+            else Cstruct.append t.linger data
+        ) application_data ;
+      drain_handshake t
 
   let reneg ?authenticator ?acceptable_cas ?cert ?(drop = true) t =
-    match t.state with
-    | `Error err  -> raise err
-    | `Eof        -> raise End_of_file
-    | `Active tls ->
-        match Tls.Engine.reneg ?authenticator ?acceptable_cas ?cert tls with
-        | None -> invalid_arg "tls: can't renegotiate"
-        | Some (tls', buf) ->
-           if drop then t.linger <- None ;
-           t.state <- `Active tls' ;
-           write_t t buf;
-           ignore (drain_handshake t : t)
+    match Tls.Engine.reneg ?authenticator ?acceptable_cas ?cert t.tls with
+    | None -> invalid_arg "tls: can't renegotiate"
+    | Some (tls', buf) ->
+      if drop then t.linger <- Cstruct.empty ;
+      t.tls <- tls' ;
+      write_t t buf ;
+      ignore (drain_handshake t : t)
 
   let key_update ?request t =
-    match t.state with
-    | `Error err  -> raise err
-    | `Eof        -> raise End_of_file
-    | `Active tls ->
-      match Tls.Engine.key_update ?request tls with
-      | Error _ -> invalid_arg "tls: can't update key"
-      | Ok (tls', buf) ->
-        t.state <- `Active tls' ;
-        write_t t buf
+    match Tls.Engine.key_update ?request t.tls with
+    | Error _ -> invalid_arg "tls: can't update key"
+    | Ok (tls', buf) ->
+      t.tls <- tls' ;
+      write_t t buf
 
   let close_tls t =
-    match t.state with
-    | `Active tls ->
-        let (_, buf) = Tls.Engine.send_close_notify tls in
-        t.state <- `Eof ;       (* XXX: this looks wrong - we're only trying to close the sending side *)
-        write_t t buf
-    | _ -> ()
-
-  (* Not sure if we need to keep both directions open on the underlying flow when closing
-     one direction at the TLS level. *)
+    let (tls, buf) = Tls.Engine.send_close_notify t.tls in
+    write_t t buf ;
+    t.tls <- tls
+
   let shutdown t = function
-    | `Send -> close_tls t
-    | `All -> close_tls t; Flow.shutdown t.flow `All
-    | `Receive -> ()  (* Not obvious how to do this with TLS, so ignore for now. *)
+    | `Send -> close_tls t ; Flow.shutdown t.flow `Send
+    | `All -> close_tls t ; Flow.shutdown t.flow `All
+    | `Receive -> close_tls t ; Flow.shutdown t.flow `Receive
```

Comment:
Not sure about this one but I have added a corresponding close on the `t.flow` believing this version to be more semantically correct than the original. However, please advise if this is not the case.